### PR TITLE
AK: Eliminate superfluous copies in Array min/max

### PR DESCRIPTION
**Pull request by tryfinally**
Array::max and Array::min make unneded copies while iterating over the array to find the extreme value. Implement without copies.

### CODE REVIEW

**Comment by Hendiadyoin1** [comment]
Your commit message has a superfluous `e` at the beginning

To fix this, use `git commit --amend` and `git push -f`, no need to make a new PR for that, fixing up singlencommits works similarly,
For more complex git stuff i recommend learning about `git rebase`

Here is a good resource on the topic:
https://youtu.be/ElRzTuYln0M

**Comment by tryfinally** [comment]
Thanks @Hendiadyoin1 for spotting that, it was actually the PR message, I edited that.

**Review by ADKaster — COMMENTED** [review]
This looks like a pretty obvious change to me! Although... could you add a couple test cases to TestArray.cpp if it's not too much trouble?
https://github.com/SerenityOS/serenity/blob/b01fcccbbd43abea157143ebfa64976dc389278c/Tests/AK/TestArray.cpp

File: AK/Array.h [hunk 1]
```diff
@@ -103,21 +103,25 @@ struct Array {
     {
         static_assert(Size > 0, "No values to max() over");
 
-        T value = __data[0];
+        size_t candidate = 0;
         for (size_t i = 1; i < Size; ++i)
```
Comment (LucasChollet) [inline AK/Array.h:107]: Can you put curly braces around both `for` body?

Comment (tryfinally) [reply AK/Array.h:107]: I can, but I got the impression it's not like that in the codebase.

Comment (LucasChollet) [reply AK/Array.h:107]: https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#omission-of-curly-braces-from-statement-blocks

**Comment by MacDue** [comment]
Small comment (and not related to your change directly), but is there a reason min()/max() return a `T` and not a `T const & ` or `T&`. It seems a little inconsistent with other APIs (like first()/last()), and would be an extra copy for larger objects.

**Comment by tryfinally** [comment]
> Small comment (and not related to your change directly), but is there a reason min()/max() return a `T` and not a `T const & ` or `T&`. It seems a little inconsistent with other APIs (like first()/last()), and would be an extra copy for larger objects.

It's a deviation for the standard to avoid dangling ref in calls like :
```cpp
min(f(), g());
max(3, 5);
```
I have created PR #18866 that fixes this and returns T or T& or T const& according to type of category of arguments.

File: AK/Array.h [hunk 2]
```diff
@@ -104,9 +104,11 @@ struct Array {
         static_assert(Size > 0, "No values to max() over");
 
         size_t candidate = 0;
-        for (size_t i = 1; i < Size; ++i)
-            if (__data[candidate] < __data[i])
+        for (size_t i = 1; i < Size; ++i) {
+            if (__data[candidate] < __data[i]) {
```
Comment (LucasChollet) [inline AK/Array.h:108]: FYI, they are not mandatory around the `if` body

**Review by LucasChollet — COMMENTED** [review]
Please rewrite the git history fixup the two first commits

**Comment by MacDue** [comment]
I was referring specifically to the Array::min()/max() not the general helpers, and I don't think there's a concern with dangling references there?

File: Tests/AK/TestArray.cpp [hunk 3]
```diff
@@ -46,3 +46,15 @@ TEST_CASE(first_index_of)
     EXPECT(array.first_index_of(7) == 7u);
     EXPECT(!array.first_index_of(42).has_value());
 }
+
+TEST_CASE(max)
+{
+    constexpr Array<int, 8> array = { 0, 31, 2, 3, 41, 5, -6, 7 };
+    EXPECT(array.max() == 41);
```
Comment (BenWiederhake2) [inline Tests/AK/TestArray.cpp:53]: If returning a reference is the entire point of this pr, then I'd like to ask you to test that, too.

Some ideas:
```
array.max() = -42;
VERIFY(array[4] == -42);
```

File: Tests/AK/TestArray.cpp [hunk 4]
```diff
@@ -46,3 +46,15 @@ TEST_CASE(first_index_of)
...
+}
+
+TEST_CASE(min)
+{
+    constexpr Array<int, 8> array = { 0, 31, 2, 3, 41, 5, -6, 7 };
+    EXPECT(array.min() == -6);
```
Comment (BenWiederhake2) [inline Tests/AK/TestArray.cpp:59]: Same here.

**Review by BenWiederhake2 — CHANGES_REQUESTED** [review]
Changing Array::min/max to return references seems like a good idea! I don't share the concerns about standard compatibility, since `AK::Array` is nonstandard anyway, and there cannot be any dangling references in this case.

However, please squash the fixup commit, and actually test the functionality you implemented.

If you'd like some pointers on how to do this:
- use `git rebase -i origin/master` to start an interactive rebase.
- in the text editor (you can change which one is selected), change the second commit to "f"
- quit the editor, done.
- i also recommend using a visualization tool like gitk to wrap your head around how the commit graph looks like.

**Pull request closed without merging**